python/runtime/cudaq/algorithms/py_translate.cpp

```diff
@@ -0,0 +1,128 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+#include "cudaq/Optimizer/CodeGen/OpenQASMEmitter.h"
+#include "cudaq/Optimizer/CodeGen/Pipelines.h"
+#include "cudaq/algorithms/draw.h" // TODO  translate.h
+#include "utils/OpaqueArguments.h"
+#include "mlir/Bindings/Python/PybindAdaptors.h"
+
+#include <iostream>
+#include <pybind11/complex.h>
+#include <pybind11/stl.h>
+
+namespace cudaq {
+
+void pyAltLaunchKernel(const std::string &, MlirModule, OpaqueArguments &,
+                       const std::vector<std::string> &);
+
+std::string getQIRLL(const std::string &name, MlirModule module,
+                     cudaq::OpaqueArguments &runtimeArgs, std::string &profile);
+
+std::string getASM(const std::string &name, MlirModule module,
+                   cudaq::OpaqueArguments &runtimeArgs);
+
+/// @brief Run `cudaq::translate` on the provided kernel.
+std::string pyTranslate(py::object &kernel, py::args args, std::string format) {
+
+  if (py::hasattr(kernel, "compile"))
+    kernel.attr("compile")();
+
+  auto name = kernel.attr("name").cast<std::string>();
+  auto module = kernel.attr("module").cast<MlirModule>();
+
+  auto result =
+      llvm::StringSwitch<std::function<std::string()>>(format)
+          .Case("qir",
+                [&]() {
+                  cudaq::OpaqueArguments args;
+                  std::string profile = "";
+                  return getQIRLL(name, module, args, profile);
+                })
+          .Cases("qir-adaptive", "qir-base",
+                 [&]() {
+                   cudaq::OpaqueArguments args;
+                   return getQIRLL(name, module, args, format);
+                 })
+          .Case("openqasm",
+                [&]() {
+                  if (py::hasattr(kernel, "arguments") &&
+                      py::len(kernel.attr("arguments")) > 0) {
+                    throw std::runtime_error("Cannot translate function with "
+                                             "arguments to OpenQasm 2.0.");
+                  }
+                  cudaq::OpaqueArguments args;
+                  return getASM(name, module, args);
+                })
+          .Default([&]() {
+            throw std::runtime_error("Invalid format to translate to: " +
+                                     format);
+            return "Failed to translate to " + format;
+          })();
+
+  return result;
+}
+
+/// @brief Bind the draw cudaq function
+void bindPyTranslate(py::module &mod) {
```

Comment:
Is this just a big comment with the final string argument? Do we really need C++ code and such a comment at runtime?

Reply:
We use quite verbose descriptions in other python bindings, so I added one here as well. `cudaq.translate` API binds to this one, so I presume the description is destined to show on hover over in the editor UI at some point (thought it does not yet)

Reply:
I shortened the description by removing most of the `qir` code in the example.